homeassistant/components/elkm1/config_flow.py

```diff
@@ -0,0 +1,164 @@
+    url = _make_url_from_data(data)
+    requires_password = url.startswith("elks://")
+
+    if requires_password and (not userid or not password):
+        raise InvalidAuth
+
+    elk = elkm1.Elk(
+        {"url": url, "userid": userid, "password": password, "element_list": ["panel"]}
+    )
+    elk.connect()
+
+    timed_out = False
+    if not await async_wait_for_elk_to_sync(elk, VALIDATE_TIMEOUT):
+        _LOGGER.error(
+            "Timed out after %d seconds while trying to sync with elkm1",
+            VALIDATE_TIMEOUT,
+        )
+        timed_out = True
+
+    elk.disconnect()
+
+    if timed_out:
+        raise CannotConnect
+    if elk.invalid_auth:
+        raise InvalidAuth
+
+    device_name = data[CONF_PREFIX] if data[CONF_PREFIX] else "ElkM1"
+    # Return info that you want to store in the config entry.
+    return {"title": device_name, CONF_HOST: url, CONF_PREFIX: slugify(prefix)}
+
+
+def _make_url_from_data(data):
+    host = data.get(CONF_HOST)
+    if host:
+        return host
+
+    protocol = PROTOCOL_MAP[data[CONF_PROTOCOL]]
+    address = data[CONF_ADDRESS]
+    return f"{protocol}{address}"
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Elk-M1 Control."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    def __init__(self):
+        """Initialize the elkm1 config flow."""
+        self.importing = False
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+            if self._url_already_configured(_make_url_from_data(user_input)):
+                return self.async_abort(reason="address_already_configured")
+
+            try:
+                info = await validate_input(user_input)
+
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+            if "base" not in errors:
+                await self.async_set_unique_id(user_input[CONF_PREFIX])
+                self._abort_if_unique_id_configured()
+
+                if self.importing:
+                    return self.async_create_entry(title=info["title"], data=user_input)
+
+                return self.async_create_entry(
+                    title=info["title"],
+                    data={
+                        CONF_HOST: info[CONF_HOST],
+                        CONF_USERNAME: user_input[CONF_USERNAME],
+                        CONF_PASSWORD: user_input[CONF_PASSWORD],
+                        CONF_AUTO_CONFIGURE: True,
+                        CONF_TEMPERATURE_UNIT: user_input[CONF_TEMPERATURE_UNIT],
+                        CONF_PREFIX: info[CONF_PREFIX],
```

Comment:
Can the prefix be automatically selected instead?

Reply:
Unfortunately the device won't give up the serial number as its used as a security "feature"
https://www.elkproducts.com/_literature_63632/M1_RS232_ASCII_Protocol